### PR TITLE
Fix 2 memory leak for #141

### PR DESCRIPTION
**Pull request by bertliao**
Fix 2 memory leak for #141

### CODE REVIEW

File: src/reqs.c [hunk 1]
```diff
@@ -681,7 +681,11 @@ static int get_all_headers (int fd, hashmap_t hashofheaders)
                 /*
                  * Append the new line to the current header field.
                  */
-                tmp = (char *) saferealloc (header, len + linelen);
+                if( NULL == header ) {
```
Comment (rofl0r) [inline src/reqs.c]: what's the point of this check ?

according to `man 3p realloc`

> The  realloc() function shall deallocate the old object pointed to by ptr and return a pointer to a new object that has the size specified by size.
> ...
> If  ptr  is a null pointer, realloc() shall be equivalent to malloc() for the specified size.

File: src/transparent-proxy.c [hunk 2]
```diff
@@ -38,17 +38,21 @@
...
         assert (url != NULL);
         assert (host != NULL);
         assert (port > 0 && port < 32768);
         assert (path != NULL);
 
         len = strlen (host) + strlen (path) + 14;
-        *url = (char *) safemalloc (len);
-        if (*url == NULL)
+        
+        tmp = (char *) saferealloc (*url, len);
```
Comment (rofl0r) [inline src/transparent-proxy.c]: what's the purpose of letting the allocation go through a temporary var ?

File: src/reqs.c (hunk 1 above)
Comment (bertliao) [reply src/reqs.c]: I found this in 1.8.4, it use safemalloc.... this bug is already fixed in the mater

File: src/transparent-proxy.c (hunk 2 above)
Comment (bertliao) [reply src/transparent-proxy.c]: The build_url is called by do_transparent_proxy/process_request, the url param is malloc in process_request. If we use safemalloc, there will be memory leak.

Comment (rofl0r) [reply src/transparent-proxy.c]: that's not really answering my question.
please try again.

also please state which OS you use, which LIBC and which architecture.

Comment (bertliao) [reply src/transparent-proxy.c]: The temporary var is not necessary
`*url = (char *) saferealloc (*url, len);`

Env:
Openwrt/Linux 3.3.8, uClibc-0.9.33.2, MIPS/x86

Comment (rofl0r) [reply src/transparent-proxy.c]: thanks. can you show the output from valgrind or whatever tool that showed you this leak ?

Comment (bertliao) [reply src/transparent-proxy.c]: Mmm, we are not using tools, we found it casually when we executed `ps` and found the process memory increased persistent. Then we limit the process in config file
`maxclients 1`
`maxspareservers 1`
`maxrequestsperchild 10000`
Start
`admin@admin:/tmp# ps | grep tinyproxy`
` 2274 admin     1088 S    /usr/sbin/tinyproxy -C /etc/tiny.conf`
` 2291 nobody    1088 S    /usr/sbin/tinyproxy -C /etc/tiny.conf`
After we accessed some web pages, it inceased slowly
`admin@admin:/tmp# ps | grep tinyproxy`
` 2274 admin     1088 S    /usr/sbin/tinyproxy -C /etc/tiny.conf`
` 2291 nobody    1176 S    /usr/sbin/tinyproxy -C /etc/tiny.conf`

At last we enabled the debugging code through NDEBUG, found the leaking code...

Comment (rofl0r) [reply src/transparent-proxy.c]: OK. i believe this 2nd leak is legit; in the worst case if you're wrong nothing will break by changing the malloc here to realloc. could you make a new PR based on master with just that change ?

Comment (bertliao) [reply src/transparent-proxy.c]: OK, plz close this PR, thanks.

**Pull request closed without merging**